VisualPinball.Unity/VisualPinball.Unity.Editor/Editors/DragPointsEditor.cs

```diff
@@ -1,522 +1,650 @@
+			RemapControlPoints(dpeditable);
+		}
+
+		public void RemoveDragPoint(int controlId)
+		{
+			IDragPointsEditable dpeditable = _target as IDragPointsEditable;
+			if (dpeditable == null){
+				return;
+			}
+
+			var idx = _controlPoints.FindIndex(cpoint => cpoint.ControlId == controlId);
+			if (idx >= 0){
+				bool removalOK = !_controlPoints[idx].DragPoint.IsLocked;
+				if (!removalOK){
+					removalOK = EditorUtility.DisplayDialog("Locked DragPoint Removal", "This Dragpoint is Locked !!\nAre you really sure you want to remove it ?", "Yes", "No");
+				}
+
+				if (removalOK){
+					PrepareUndo("Removing Drag Point");
+					List<DragPointData> dpoints = new List<DragPointData>(dpeditable.GetDragPoints());
+					dpoints.RemoveAt(idx);
+					dpeditable.SetDragPoints(dpoints.ToArray());
+					_controlPoints.RemoveAt(idx);
+					RemapControlPoints(dpeditable);
+				}
+			}
+		}
+
+		private void ClearAllSelection()
+		{
+			foreach (var cpoint in _controlPoints)
+			{
+				cpoint.IsSelected = false;
+			}
+		}
+
+		private void UpdateDragPointsLock()
+		{
+			IEditableItemBehavior editable = _target as IEditableItemBehavior;
+			bool lockChange = false;
+			foreach (var cpoint in _controlPoints)
+			{
+				if (cpoint.DragPoint.IsLocked != editable.IsLocked)
+				{
+					cpoint.DragPoint.IsLocked = editable.IsLocked;
+					lockChange = true;
+				}
+			}
+			if (lockChange)
+			{
+				SceneView.RepaintAll();
+			}
+		}
+
+		private void OnDragPointPositionChange(Vector3 newPos, params (string,object)[] plist)
 		{
-			foreach (var cpoint in _controlPoints)
-			{
-				cpoint.IsSelected = false;
-			}
-		}
-
-		public void OnSceneGUI(Object target)
-		{
-			_target = target;
-			IEditableItemBehavior editable = _target as IEditableItemBehavior;
-			IDragPointsEditable dpeditable = _target as IDragPointsEditable;
-			Behaviour bh = _target as Behaviour;
-
-			if (bh == null || dpeditable == null || !dpeditable.DragPointEditEnabled){
-				return;
-			}
-
-			if (_controlPoints.Count != dpeditable.GetDragPoints().Length){
-				RebuildControlPoints(dpeditable);
-			}
-
-			Vector3 offset = dpeditable.GetEditableOffset();
-			Matrix4x4 lwMat = bh.transform.localToWorldMatrix;
-			Matrix4x4 wlMat = bh.transform.worldToLocalMatrix;
-
-			switch (Event.current.type)
-			{
-				case EventType.Layout:
+			Matrix4x4 wlMat = Enumerable.Count<(string, object)>(plist, pair => pair.Item1 == "wlMat") > 0 ? (Matrix4x4)Enumerable.First<(string, object)>(plist, pair => pair.Item1 == "wlMat").Item2 : Matrix4x4.identity;
+			Vector3 offset = Enumerable.Count<(string, object)>(plist, pair => pair.Item1 == "offset") > 0 ? (Vector3)Enumerable.First<(string, object)>(plist, pair => pair.Item1 == "offset").Item2 : Vector3.zero;
+			IDragPointsEditable dpeditable = Enumerable.Count<(string, object)>(plist, pair => pair.Item1 == "dpeditable") > 0 ? (IDragPointsEditable)Enumerable.First<(string, object)>(plist, pair => pair.Item1 == "dpeditable").Item2 : null;

```

Comment:
This is incomprehensible for me ;)

Tuples that match by string? Casts to matrix and vector? Does this have to be so generic, or can we get that `plist` better typed somehow?

Reply:
Hey realize these where useless as params, i can find them back from inside the delegate. Will remove.